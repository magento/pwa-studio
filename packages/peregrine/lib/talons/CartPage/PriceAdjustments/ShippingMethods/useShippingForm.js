import { useCallback, useEffect, useMemo } from 'react';
import { useApolloClient, useMutation } from '@apollo/client';

import { useCartContext } from '../../../../context/cart';
import mergeOperations from '@magento/peregrine/lib/util/shallowMerge';
import DEFAULT_OPERATIONS from './shippingMethods.gql';

/**
 * GraphQL currently requires a complete address before it will return
 * estimated shipping prices, even though it only needs Country, State,
 * and Zip. Assuming this is a bug or oversight, we're going to mock the
 * data points we don't want to bother collecting from the Customer at this
 * step in the process. We need to be very mindful that these values are never
 * displayed to the user.
 *
 * @ignore
 */
export const MOCKED_ADDRESS = {
    city: 'city',
    firstname: 'firstname',
    lastname: 'lastname',
    street: ['street'],
    telephone: 'telephone'
};

/**
 * Contains logic for a shipping form component.
 * It performs effects and returns props data related to rendering a shipping address form.
 *
 * This talon performs the following effects:
 *
 * - Manage the updating state of the cart while the mutation for setting the shipping address is in flight
 *
 * @function
 *
 * @param {Object} props
 * @param {SelectShippingFields} props.selectedValues The values from the select input fields in the shipping form
 * @param {function} props.setIsCartUpdating Callback function for setting the update state for the cart.
 * @param {ShippingFormMutations} props.mutations GraphQL mutations for the shipping form.
 * @param {ShippingFormQueries} props.queries GraphQL queries for the shipping form.
 *
 * @returns {ShippingFormTalonProps}
 *
 * @example <caption>Importing into your project</caption>
 * import { useShippingForm } from '@magento/peregrine/lib/talons/CartPage/PriceAdjustments/ShippingMethods/useShippingForm';
 */
export const useShippingForm = props => {
    const operations = mergeOperations(DEFAULT_OPERATIONS, props.operations);
    const { setShippingAddressMutation, getShippingMethodsQuery } = operations;
    const { selectedValues, setIsCartUpdating } = props;

    const [{ cartId }] = useCartContext();
    const apolloClient = useApolloClient();

    const [
        setShippingAddress,
        {
            called: isSetShippingAddressCalled,
            error: errorSettingShippingAddress,
            loading: isSetShippingLoading
        }
    ] = useMutation(setShippingAddressMutation);

    useEffect(() => {
        if (isSetShippingAddressCalled) {
            setIsCartUpdating(isSetShippingLoading);
        }
    }, [isSetShippingLoading, isSetShippingAddressCalled, setIsCartUpdating]);

    /**
     * @ignore
     *
     * When the zip value is changed, go ahead and manually wipe out that
     * portion of the cache, which will cause the components subscribed to
     * that state to re-render. Ideally we would send a mutation to clear the
     * shipping address set on the cart, to keep those states in sync, but the
     * GraphQL API does not currently supported clearing addresses.
     */
    const handleZipChange = useCallback(
        zip => {
            if (zip !== selectedValues.zip) {
                const data = apolloClient.readQuery({
                    query: getShippingMethodsQuery,
                    variables: {
                        cartId
                    }
                });

                const { cart } = data;
                const { shipping_addresses: shippingAddresses } = cart;

                if (shippingAddresses.length) {
                    const primaryAddress = shippingAddresses[0];
                    const {
                        available_shipping_methods: availableMethods
                    } = primaryAddress;
                    if (availableMethods.length) {
                        apolloClient.writeQuery({
                            query: getShippingMethodsQuery,
                            data: {
                                cart: {
                                    ...cart,
                                    shipping_addresses: [
                                        {
                                            // TODO: we're losing addresses here, we only ever put the first one back in
                                            ...primaryAddress,
                                            available_shipping_methods: []
                                        }
                                    ]
                                }
                            }
                        });
                    }
                }
            }
        },
        [apolloClient, cartId, selectedValues.zip, getShippingMethodsQuery]
    );

    const handleOnSubmit = useCallback(
        formValues => {
            const { country, region, zip } = formValues;
            if (country && region && zip) {
                setShippingAddress({
                    variables: {
                        cartId,
                        address: {
                            ...MOCKED_ADDRESS,
                            country_code: country,
                            postcode: zip,
                            region
                        }
                    }
                });
            }
        },
        [cartId, setShippingAddress]
    );

    const errors = useMemo(
        () =>
            new Map([
                ['setShippingAddressMutation', errorSettingShippingAddress]
            ]),
        [errorSettingShippingAddress]
    );

    return {
        errors,
        handleOnSubmit,
        handleZipChange,
        isSetShippingLoading
    };
};

/** JSDocs type definitions */

/**
 * Values for the select input fields on the shipping form.
 * This is a prop used by the {@link useShippingForm} talon.
 *
 * @typedef {Object} SelectShippingFields
 *
 * @property {String} country Country shipping destination
 * @property {String} region Country's region shipping destination
 * @property {String} zip Country's zip code shipping destination
 */

/**
 * GraphQL mutations for the shipping form.
 * This is a type used by the {@link useShippingForm} talon.
 *
 * @typedef {Object} ShippingFormMutations
 *
 * @property {GraphQLAST} setShippingAddressMutation Mutation for setting the shipping address on a cart
 *
 * @see [shippingForm.js]{@link https://github.com/magento/pwa-studio/blob/develop/packages/venia-ui/lib/components/CartPage/PriceAdjustments/ShippingMethods/shippingForm.js}
 * for the query used in Venia
 */

/**
 * GraphQL queries for the shipping form.
 * This is a type used by the {@link useShippingForm} talon.
 *
 * @typedef {Object} ShippingFormQueries
 *
 * @property {GraphQLAST} shippingMethodsQuery Query for getting data about available shipping methods
 *
 * @see [shippingMethods.gql.js]{@link https://github.com/magento/pwa-studio/blob/develop/packages/venia-ui/lib/components/CartPage/PriceAdjustments/ShippingMethods/shippingMethods.gql.js}
 * for the query used in Venia
 */

/**
 * Object type returned by the {@link useShippingForm} talon.
 * It provides props data to use when rendering a shipping form component.
 *
 * @typedef {Object} ShippingFormTalonProps
 *
 * @property {Array<Error>} formErrors A list of form errors
 * @property {function} handleOnSubmit Callback function to handle form submissions
 * @property {function} handleZipChange Callback function to handle a zip code change
 * @property {boolean} isSetShippingLoading True if the cart shipping information is being set. False otherwise.
 */
