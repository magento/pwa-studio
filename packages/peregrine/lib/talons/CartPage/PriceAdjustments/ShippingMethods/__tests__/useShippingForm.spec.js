import React from 'react';
import { useApolloClient, useMutation } from '@apollo/client';

import createTestInstance from '../../../../../util/createTestInstance';
import { MOCKED_ADDRESS, useShippingForm } from '../useShippingForm';
import { act } from 'react-test-renderer';

jest.mock('@apollo/client', () => ({
    useApolloClient: jest.fn(),
    useMutation: jest.fn().mockReturnValue([
        jest.fn(),
        {
            called: false,
            error: undefined,
            loading: false
        }
    ])
}));

jest.mock('../../../../../context/cart', () => {
    const state = {
        cartId: 'cart123'
    };
    const api = {};
    const useCartContext = jest.fn(() => [state, api]);

    return { useCartContext };
});

jest.mock('../shippingMethods.gql', () => ({
    getShippingMethodsQuery: 'getShippingMethodsQuery',
    setShippingAddressMutation: 'setShippingAddressMutation'
}));

const Component = props => {
    const talonProps = useShippingForm(props);
    return <i talonProps={talonProps} />;
};

const props = {
    selectedValues: {
        zip: '78725'
    },
    setIsCartUpdating: jest.fn(),
    mutations: {},
    queries: {
        shippingMethodsQuery: 'shippingMethodsQuery'
    }
};

test('returns correct shape', () => {
    const tree = createTestInstance(<Component {...props} />);
    const { root } = tree;
    const { talonProps } = root.findByType('i').props;

    expect(talonProps).toMatchSnapshot();
});

test('returns errors and loading state from Apollo', () => {
    useMutation.mockReturnValueOnce([
        jest.fn(),
        {
            called: true,
            error: 'Apollo Error',
            loading: true
        }
    ]);

    const tree = createTestInstance(<Component {...props} />);
    const { root } = tree;
    const { talonProps } = root.findByType('i').props;

    expect(talonProps).toMatchSnapshot();
});

test('handleOnSubmit() runs the mutation with the expected values', () => {
    const setShippingAddress = jest.fn();
    useMutation.mockReturnValueOnce([setShippingAddress, {}]);

    const tree = createTestInstance(<Component {...props} />);
    const { root } = tree;
    const { talonProps } = root.findByType('i').props;

    const { handleOnSubmit } = talonProps;

    const formValues = {
        country: 'USA',
        region: 'Texas',
        zip: '78725'
    };

    act(() => {
        handleOnSubmit({});
    });

    expect(setShippingAddress).not.toHaveBeenCalled();

    act(() => {
        handleOnSubmit(formValues);
    });

    expect(setShippingAddress.mock.calls[0][0]).toMatchSnapshot();
});

describe('handleZipChange()', () => {
    test('writes the correct value to the cache', () => {
        const mockReadQuery = jest.fn(() => {
            return {
                cart: {
                    shipping_addresses: [
                        {
                            ...MOCKED_ADDRESS,
                            country_code: 'USA',
                            postcode: '78725',
                            region: 'Texas',
                            available_shipping_methods: [
                                {
                                    available: true
                                }
                            ]
                        }
                    ]
                }
            };
        });
        const mockWriteQuery = jest.fn();
        useApolloClient.mockReturnValueOnce({
            readQuery: mockReadQuery,
            writeQuery: mockWriteQuery
        });
        const tree = createTestInstance(<Component {...props} />);
        const { root } = tree;
        const { talonProps } = root.findByType('i').props;

        const { handleZipChange } = talonProps;

        act(() => {
            handleZipChange('78743');
        });

        expect(mockWriteQuery.mock.calls[0][0]).toMatchSnapshot();
    });

    test('does not update cache when there are no available shipping methods', () => {
        const mockReadQuery = jest.fn(() => {
            return {
                cart: {
                    shipping_addresses: [
                        {
                            ...MOCKED_ADDRESS,
                            country_code: 'USA',
                            postcode: '78725',
                            region: 'Texas',
                            available_shipping_methods: []
                        }
                    ]
                }
            };
        });
        const mockWriteQuery = jest.fn();
        useApolloClient.mockReturnValueOnce({
            readQuery: mockReadQuery,
            writeQuery: mockWriteQuery
        });
        const tree = createTestInstance(<Component {...props} />);
        const { root } = tree;
        const { talonProps } = root.findByType('i').props;

        const { handleZipChange } = talonProps;

        act(() => {
            handleZipChange('78743');
        });

        expect(mockReadQuery).toHaveBeenCalled();
        expect(mockWriteQuery).not.toHaveBeenCalled();
    });

    test('does not update cache when there are no available shipping addresses', () => {
        const mockReadQuery = jest.fn(() => {
            return {
                cart: {
                    shipping_addresses: []
                }
            };
        });
        const mockWriteQuery = jest.fn();
        useApolloClient.mockReturnValueOnce({
            readQuery: mockReadQuery,
            writeQuery: mockWriteQuery
        });
        const tree = createTestInstance(<Component {...props} />);
        const { root } = tree;
        const { talonProps } = root.findByType('i').props;

        const { handleZipChange } = talonProps;

        act(() => {
            handleZipChange('78743');
        });

        expect(mockReadQuery).toHaveBeenCalled();
        expect(mockWriteQuery).not.toHaveBeenCalled();
    });

    test('does not read query value from apollo when changed zip is the same as selected', () => {
        const mockReadQuery = jest.fn();
        const mockWriteQuery = jest.fn();
        useApolloClient.mockReturnValueOnce({
            readQuery: mockReadQuery,
            writeQuery: mockWriteQuery
        });
        const tree = createTestInstance(<Component {...props} />);
        const { root } = tree;
        const { talonProps } = root.findByType('i').props;

        const { handleZipChange } = talonProps;

        act(() => {
            handleZipChange('78725');
        });

        expect(mockReadQuery).not.toHaveBeenCalled();
        expect(mockWriteQuery).not.toHaveBeenCalled();
    });
});
