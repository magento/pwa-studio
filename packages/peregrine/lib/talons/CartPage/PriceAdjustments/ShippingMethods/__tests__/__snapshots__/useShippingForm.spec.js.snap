// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleOnSubmit() runs the mutation with the expected values 1`] = `
Object {
  "variables": Object {
    "address": Object {
      "city": "city",
      "country_code": "USA",
      "firstname": "firstname",
      "lastname": "lastname",
      "postcode": "78725",
      "region": "Texas",
      "street": Array [
        "street",
      ],
      "telephone": "telephone",
    },
    "cartId": "cart123",
  },
}
`;

exports[`handleZipChange() writes the correct value to the cache 1`] = `
Object {
  "data": Object {
    "cart": Object {
      "shipping_addresses": Array [
        Object {
          "available_shipping_methods": Array [],
          "city": "city",
          "country_code": "USA",
          "firstname": "firstname",
          "lastname": "lastname",
          "postcode": "78725",
          "region": "Texas",
          "street": Array [
            "street",
          ],
          "telephone": "telephone",
        },
      ],
    },
  },
  "query": "getShippingMethodsQuery",
}
`;

exports[`returns correct shape 1`] = `
Object {
  "errors": Map {
    "setShippingAddressMutation" => undefined,
  },
  "handleOnSubmit": [Function],
  "handleZipChange": [Function],
  "isSetShippingLoading": false,
}
`;

exports[`returns errors and loading state from Apollo 1`] = `
Object {
  "errors": Map {
    "setShippingAddressMutation" => "Apollo Error",
  },
  "handleOnSubmit": [Function],
  "handleZipChange": [Function],
  "isSetShippingLoading": true,
}
`;
